protocols/autonat/src/v2/protocol.rs

```diff
@@ -0,0 +1,362 @@
+                let addr_idx = ok_or_invalid_data!(addrIdx)? as usize;
+                let num_bytes = ok_or_invalid_data!(numBytes)? as usize;
+                Ok(Self::Data(DialDataRequest {
+                    addr_idx,
+                    num_bytes,
+                }))
+            }
+            _ => Err(new_io_invalid_data_err(
+                "invalid message type, expected dialResponse or dialDataRequest",
+            )),
+        }
+    }
+}
+
+impl From<Response> for proto::Message {
+    fn from(val: Response) -> Self {
+        match val {
+            Response::Dial(DialResponse {
+                status,
+                addr_idx,
+                dial_status,
+            }) => proto::Message {
+                msg: proto::mod_Message::OneOfmsg::dialResponse(proto::DialResponse {
+                    status: Some(status),
+                    addrIdx: Some(addr_idx as u32),
+                    dialStatus: Some(dial_status),
+                }),
+            },
+            Response::Data(DialDataRequest {
+                addr_idx,
+                num_bytes,
+            }) => proto::Message {
+                msg: proto::mod_Message::OneOfmsg::dialDataRequest(proto::DialDataRequest {
+                    addrIdx: Some(addr_idx as u32),
+                    numBytes: Some(num_bytes as u64),
+                }),
+            },
+        }
+    }
+}
+
+impl DialDataRequest {
+    pub(crate) fn from_rng<R: rand_core::RngCore>(addr_idx: usize, mut rng: R) -> Self {
+        let num_bytes = rng.gen_range(DATA_LEN_LOWER_BOUND..=DATA_LEN_UPPER_BOUND);
+        Self {
+            addr_idx,
+            num_bytes,
+        }
+    }
+}
+
+const DIAL_BACK_MAX_SIZE: usize = 10;
+
+pub(crate) async fn dial_back(stream: impl AsyncWrite + Unpin, nonce: Nonce) -> io::Result<()> {
+    let msg = proto::DialBack { nonce: Some(nonce) };
+    let mut framed = FramedWrite::new(stream, Codec::<proto::DialBack>::new(DIAL_BACK_MAX_SIZE));
+
+    framed
+        .send(msg)
+        .await
+        .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+
+    Ok(())
+}
+
+pub(crate) async fn recv_dial_back(stream: impl AsyncRead + Unpin) -> io::Result<Nonce> {
+    let framed = &mut FramedRead::new(stream, Codec::<proto::DialBack>::new(DIAL_BACK_MAX_SIZE));
+    let proto::DialBack { nonce } = framed
+        .next()
+        .await
+        .ok_or(io::Error::from(io::ErrorKind::UnexpectedEof))??;
+    let nonce = ok_or_invalid_data!(nonce)?;
+
+    Ok(nonce)
+}
+
+pub(crate) async fn dial_back_response(stream: impl AsyncWrite + Unpin) -> io::Result<()> {
+    let msg = proto::DialBackResponse {
+        status: Some(proto::mod_DialBackResponse::DialBackStatus::OK),
+    };
+    let mut framed = FramedWrite::new(
+        stream,
+        Codec::<proto::DialBackResponse>::new(DIAL_BACK_MAX_SIZE),
+    );
+    framed
+        .send(msg)
+        .await
+        .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+    framed.close().await?;
```

Comment:
We observed that the `close()` may fail if the client has already closed the stream
Suggestion: do not close at all (the stream will be closed when the stream is dropped)

In identify prococol , we have the same behaviour
https://github.com/libp2p/rust-libp2p/blob/ae0a187e4adf8946e9046c001154b364ca637ade/protocols/identify/src/protocol.rs#L147

Reply:
Yes ! No more errors related to stream closed too early